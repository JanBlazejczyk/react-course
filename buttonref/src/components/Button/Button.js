import { forwardRef } from "react";

const styles = {
  backgroundColor: "blue",
  padding: 20
};

const Button = forwardRef((props, ref) => {
  return (
    <button style={styles} ref={ref} {...props}>
      I change color
    </button>
  );
});

export default Button;
