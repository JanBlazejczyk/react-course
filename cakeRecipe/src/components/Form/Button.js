import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

function Button({ children, bgColor, color, icon }) {
  return (
    <button
      style={{
        // props
        color: `${color}`,
        backgroundColor: `${bgColor}`,
        // own styles
        paddingTop: 10,
        paddingBottom: 10,
        paddingLeft: 15,
        paddingRight: 15,
        borderRadius: "1rem",
        borderWidth: "0px",
        // border: 0.1,
        fontWeight: 1000,
        opacity: 0.8,
        margin: "5px"
      }}
    >
      <FontAwesomeIcon icon={icon} size="xs" />
      {children}
    </button>
  );
}

Button.defaultProps = {
  bgColor: "#FFCAD4",
  color: "#293241"
};

export default Button;
