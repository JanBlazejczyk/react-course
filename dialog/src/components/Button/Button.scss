.button {
  padding: 15px;
  font-weight: 800;
  cursor: pointer;
  border-radius: 0.5em;
  border: none;
  margin: 5px;

  &--confirm {
    background-color: green;
  }
  &--abort {
    background-color: red;
  }
  &--close {
    background-color: yellow;
  }
}
