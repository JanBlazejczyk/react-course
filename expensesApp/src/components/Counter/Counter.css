.counter__container {
  border: solid;
  height: 15%;
  margin: 1rem 0.5rem 2rem 0.5rem;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  color: #3d405b;
  background-color: #f4f1de;
  border: 1px solid #3d405b;
  border-radius: 0.3em;
  font-weight: 500;
  padding: 0.5rem;
}

.counter__title {
  margin-top: 0;
  margin-bottom: 0.5rem;
}

.counter {
  font-weight: 800;
}

.counter__red {
  font-weight: 800;
  color: red;
}
