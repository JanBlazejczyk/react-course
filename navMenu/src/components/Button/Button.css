.button {
  height: 30px;
  padding: 5px;
  background-color: transparent;
  border: none;
  cursor: pointer;
}
