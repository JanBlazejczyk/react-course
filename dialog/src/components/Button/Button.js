import "./Button.scss";

import classNames from "classnames";

function Button({ children, confirm, abort, close, handleClick }) {
  const classes = classNames({
    button: true,
    "button--confirm": confirm,
    "button--abort": abort,
    "button--close": close
  });

  return (
    <button className={classes} onClick={handleClick}>
      {children}
    </button>
  );
}

export default Button;
