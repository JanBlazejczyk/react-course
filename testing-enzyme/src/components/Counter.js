import { useState, useEffect } from "react";

import Button from "./Button";

function Counter({ start = 0 }) {
  const [counter, setCounter] = useState(0);
  const [initCounterValue, setInitCounterValue] = useState(start);
  const [inputValue, setInputValue] = useState(0);

  const handleIncrement = () => {
    setCounter(Number(counter) + 1);
  }

  const handleDecrement = () => {
    setCounter(counter - 1);
  }

  const handleChange = () => {
    setCounter(inputValue);
    setInitCounterValue(inputValue);
  }

  const handleReset = () => {
    setCounter(initCounterValue);
  }

  const handleInputChange = (event) => {
    setInputValue(event.target.value);
  }

  useEffect(() => {
    setCounter(initCounterValue); // when the component is mounted the counter is set to start value
  }, [initCounterValue]);

  return (
    <div>
      <div className="counter-value">{counter}</div>
      <input className="counter-input" type="number" value={inputValue} onChange={handleInputChange} />
      <Button className="change-btn" label="Change" handleClick={handleChange} />
      <Button className="reset-btn" label="Reset" handleClick={handleReset} />
      <Button className="decrement-btn" label="-" handleClick={handleDecrement} />
      <Button className="increment-btn" label="+" handleClick={handleIncrement} />
    </div>
  )
}

export default Counter;