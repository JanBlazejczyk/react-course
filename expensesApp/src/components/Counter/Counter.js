import "./Counter.css";

function Counter({ incomeArray, expensesArray }) {
  const calculateBalance = (incomeArray, expensesArray) => {
    const balancePlusIncome = incomeArray.reduce((balance, transaction) => {
      return (balance += transaction.amount);
    }, 0);

    const finalBalance = expensesArray.reduce((balance, transaction) => {
      return (balance -= transaction.amount);
    }, balancePlusIncome);

    return (
      <span className={finalBalance >= 0 ? "counter" : "counter__red"}>
        {finalBalance}
      </span>
    );
  };

  return (
    <div className="counter__container">
      <h4 className="counter__title">Balance</h4>
      <div>{calculateBalance(incomeArray, expensesArray)}</div>
    </div>
  );
}

export default Counter;
