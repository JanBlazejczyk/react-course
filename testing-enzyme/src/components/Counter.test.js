/* eslint-disable jest/valid-expect */
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

import Counter from "./Counter";
import Button from "./Button";


describe("Counter component", () => {
  it("should render correctly", () => {
    const CounterComponent = renderer.create(<Counter />).toJSON();
    expect(CounterComponent).toMatchSnapshot();
  });

  it("should have the initial state of Counter == 0 if start is not defined", () => {
    const wrapper = shallow(<Counter />);
    const counterValue = wrapper.find(".counter-value").text();
    expect(counterValue).toEqual("0");
  });

  it("should have the initial state of the Counter == 10 if start == 10", () => {
    const wrapper = mount(<Counter start="10" />);
    expect(wrapper.prop('start')).toEqual("10");
    const counterValue = wrapper.find(".counter-value").text();
    expect(counterValue).toEqual("10");
  });

  it("should have buttons to increment and decrement the counter", () => {
    const wrapper = shallow(<Counter />);
    expect(wrapper.containsAllMatchingElements([
      <Button label="+" />,
      <Button label="-" />,
    ])).toEqual(true);
  });

  it("should decrement the counter value when the decrement button is clicked", () => {
    const wrapper = mount(<Counter />);
    const decrementButton = wrapper.find(".decrement-btn");
    let counterValue = wrapper.find(".counter-value").text();
    expect(counterValue).toEqual("0");
    decrementButton.simulate("click");
    counterValue = wrapper.find(".counter-value").text();
    expect(counterValue).toEqual("-1");
  });

  it("should increment the counter value when the increment button is clicked", () => {
    const wrapper = mount(<Counter />);
    const incrementButton = wrapper.find(".increment-btn");
    let counterValue = wrapper.find(".counter-value").text();
    expect(counterValue).toEqual("0");
    incrementButton.simulate("click");
    counterValue = wrapper.find(".counter-value").text();
    expect(counterValue).toEqual("1");
  });

  it("should change the counter state when the change button is clicked", () => {
    // value of the input field is not the value of the counter
    const wrapper = mount(<Counter />);

    // change the value in the input field to 20
    const input = wrapper.find(".counter-input");
    input.simulate('change', { target: { value: "20" } });

    // before the button is clicked the counter value is 0
    let counterValue = wrapper.find(".counter-value").text();
    expect(counterValue).toEqual("0");

    // the button is clicked
    const changeButton = wrapper.find(".change-btn");
    changeButton.simulate("click");

    // value of the input field is the value of the counter
    counterValue = wrapper.find(".counter-value").text();
    expect(counterValue).toEqual("20");
  });

  it("should have the counter value == start after reset button is clicked", () => {
    const wrapper = mount(<Counter start="10" />);
    let counterValue = wrapper.find(".counter-value").text();
    const resetButton = wrapper.find(".reset-btn");
    const incrementButton = wrapper.find(".increment-btn");

    // change the value of the counter to other than initial
    incrementButton.simulate("click");
    counterValue = wrapper.find(".counter-value").text();
    expect(counterValue).toEqual("11");

    // click the reset button and assume that it has the value of the start prop
    resetButton.simulate("click");
    counterValue = wrapper.find(".counter-value").text();
    expect(counterValue).toEqual("10");
  });
})