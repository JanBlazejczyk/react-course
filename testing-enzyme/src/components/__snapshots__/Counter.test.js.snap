// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component should render correctly 1`] = `
<div>
  <div
    className="counter-value"
  >
    0
  </div>
  <input
    className="counter-input"
    onChange={[Function]}
    type="number"
    value={0}
  />
  <button
    onClick={[Function]}
  >
    Change
  </button>
  <button
    onClick={[Function]}
  >
    Reset
  </button>
  <button
    onClick={[Function]}
  >
    -
  </button>
  <button
    onClick={[Function]}
  >
    +
  </button>
</div>
`;
